STT/Core/SttLog.swift

```diff
@@ -27,86 +27,133 @@
     }
-    open class func error(message: String, key: String) {
-        log(type: "error", message: message, key: key)
+    func emergency(_ closure: @autoclosure () -> Any?, functionName: StaticString = #function, fileName: StaticString = #file, lineNumber: Int = #line) {
+        self.log(closure, level: .emergency, functionName: String(functionName), fileName: String(fileName), lineNumber: lineNumber)
     }
-    
-    fileprivate class func log(type: String, message: String, key: String) {
-        fatalError()
+
+    // deprecated
+    @available(swift, obsoleted: 5.0)
+    var logInSystem: Bool { false }
+
+    @available(swift, obsoleted: 5.0, renamed: "verbose(_functionName:fileName:lineNumber:)")
+    func trace(message: String, key: String) { }
+    @available(swift, obsoleted: 5.0, renamed: "verbose(_functionName:fileName:lineNumber:)")
+    func warning(message: String, key: String) { }
+    @available(swift, obsoleted: 5.0, renamed: "verbose(_functionName:fileName:lineNumber:)")
+    func error(message: String, key: String) { }
+}
+
+public enum SttLogLevel: Int, CaseIterable {
+    case verbose
+    case debug
+    case info
+    case notice
+    case warning
+    case error
+    case severe // aka critical
+    case alert
+    case emergency
+    case none
+
+    public var description: String {
+        switch self {
+        case .verbose:
+            return "Verbose"
+        case .debug:
+            return "Debug"
+        case .info:
+            return "Info"
+        case .notice:
+            return "Notice"
+        case .warning:
+            return "Warning"
+        case .error:
+            return "Error"
+        case .severe:
+            return "Severe"
+        case .alert:
+            return "Alert"
+        case .emergency:
+            return "Emergency"
+        case .none:
+            return "None"
+        }
+    }
+
+    func isEnabledFor(level: SttLogLevel) -> Bool {
+        level.rawValue >= self.rawValue
     }
 }
 
-internal class SttLogDateConverter: ConverterType {
+open class SttLog: SttLogType {
+
+    public static func register(logger: SttLogType) {
+        shared = logger
+    }
+    
+    public static var shared: SttLogType = SttLog()
+
+    public var destinations: SttLogLevel = .verbose
     
-    func convert(value: Date, parameter: Any?) -> String {
-        let formatter = DateFormatter()
-        formatter.locale = Locale.current
-        formatter.timeZone = TimeZone.current
-        formatter.dateFormat = "MM-dd-yyyy HH:mm:ss:SSSS"
-        
-        return formatter.string(from: value)
+    private init() { }
+
+    public func log(_ closure: () -> Any?, level: SttLogLevel, functionName: String, fileName: String, lineNumber: Int) {
+        guard destinations.isEnabledFor(level: level) else { return }
+        guard let closureResult = closure() else { return }
```

Comment:
can be in single guard statement

 `guard destinations.isEnabledFor(level: level),
  let closureResult = closure() else { return }`